content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,83 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+- **The reference solution shouldn't be accessible to the user solution.** It should not be possible to implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global and/or public function.
+Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
+
+
+## Fixed Tests
+
+Fixed tests are tests with predetermined inputs and outputs, and do not change between test runs.
+
+- **Each requirement should have a corresponding fixed test.** Every aspect of the specification in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s).
+- **Tests should check not only valid inputs but also invalid inputs and edge cases if the problem allows them**. `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.
+- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.
+
+
+## Random Tests
+
+Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.
+
+- **Each requirement should have at least one (preferably more) corresponding random test case.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by chance.
+- **Random inputs should be _truly_ (pseudo-)random**, i.e. they should be _dynamically_ generated on every run with the help of suitable library functions such as `Math.random()` in JavaScript or `rand()` in C. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
+- In some languages such as C/C++, **make sure that the RNGs are properly seeded with the current time at the beginning** using `srand(time(NULL))` or equivalent; otherwise, the same pseudorandom sequence will be generated on every run and your "random" inputs will devolve into fixed ones.
```

Comment:
I think this part is too detailed and it rather belongs to some C howto. I added a note on proper use of random APIs above.

```suggestion
```